apps/whale-api/src/module.api/governance.service.ts

```diff
@@ -0,0 +1,160 @@
+import { ApiPagedResponse } from './_core/api.paged.response'
+import { PaginationQuery } from './_core/api.query'
+import { NotFoundApiException } from './_core/api.error'
+import {
+  GovernanceProposal,
+  GovernanceProposalStatus,
+  GovernanceProposalType,
+  ProposalMasternodeType,
+  ProposalVoteResultType,
+  ProposalVotesResult
+} from '@defichain/whale-api-client/dist/api/governance'
+
+@Injectable()
+export class GovernanceService {
+  constructor (private readonly client: JsonRpcClient) {}
+
+  async list (
+    query: PaginationQuery = {
+      size: 30
+    },
+    status?: ListProposalsStatus,
+    type?: ListProposalsType,
+    cycle?: number
+  ): Promise<ApiPagedResponse<GovernanceProposal>> {
+    const next = query.next !== undefined ? String(query.next) : undefined
+    const size = query.size > 30 ? 30 : query.size
+    const list = await this.client.governance.listGovProposals({
+      type,
+      status,
+      cycle,
+      pagination: {
+        start: next,
+        limit: size
+      }
+    })
+
+    const proposals = list.map((proposal: ProposalInfo) => {
+      return this.mapGovernanceProposal(proposal)
+    })
+
+    return ApiPagedResponse.of(proposals, size, (proposal) => {
+      return proposal.proposalId
+    })
+  }
+
+  async get (id: string): Promise<GovernanceProposal> {
+    try {
+      const proposal = await this.client.governance.getGovProposal(id)
+      return this.mapGovernanceProposal(proposal)
+    } catch (err) {
+      if (
+        err instanceof RpcApiError &&
+        (err?.payload?.message === `Proposal <${id}> not found` ||
+          err?.payload?.message ===
+            "proposalId must be of length 64 (not 3, for '999')")
+      ) {
+        throw new NotFoundApiException('Unable to find proposal')
+      } else {
+        throw new BadRequestException(err)
+      }
+    }
+  }
+
+  async getProposalVotes (
+    query: PaginationQuery = {
+      size: 30
+    },
+    id: string,
+    masternode: string | ProposalMasternodeType,
+    cycle: number
+  ): Promise<ApiPagedResponse<ProposalVotesResult>> {
+    try {
+      const next = query.next !== undefined ? Number(query.next) : undefined
+      const size = query.size > 30 ? 30 : query.size
+      const list = await this.client.governance.listGovProposalVotes({
+        proposalId: id,
+        masternode,
+        cycle,
+        pagination: {
+          start: next,
+          limit: size
+        }
+      })
+      const votes = list.map((v) => ({
+        ...v,
+        vote: mapGovernanceProposalVoteResult(v.vote)
+      }))
+
+      return ApiPagedResponse.of(votes, size, () => {
+        return (votes.length - 1).toString()
+      })
+    } catch (err) {
+      if (
+        err instanceof RpcApiError &&
+        (err?.payload?.message === `Proposal <${id}> not found` ||
+          err?.payload?.message ===
+            "proposalId must be of length 64 (not 3, for '999')")
```

Comment:
Hmm? Where is this `(not 3, for '999')` from? `ain`?

Reply:
Ops removed, that part should be dynamic, updated to use `.includes()`